### PR TITLE
Fix: flush writers before potentially expecting a response

### LINKED ISSUE #1537: rustls doesn't flush the writer
The lib gets stuck if used with, for instance, a buffered socket because it never flushes the writer.

I modified the [simple example](https://github.com/rustls/rustls/blob/v/0.21.7/examples/src/bin/simpleclient.rs) to demonstrate the issue:

```rust
use std::sync::Arc;

use std::io::{stdout, Read, Write};
use std::net::TcpStream;

use rustls::{OwnedTrustAnchor, RootCertStore};

// --- [modification] ---
use std::io::{BufReader, BufWriter, IoSlice, IoSliceMut, Result as IoResult};

#[derive(Debug)]
pub(crate) struct BufTcpStream {
    pub(super) read: BufReader<TcpStream>,
    pub(super) write: BufWriter<TcpStream>,
}

impl BufTcpStream {
    pub(super) fn new(socket: &TcpStream) -> Self {
        let read = BufReader::new(socket.try_clone().unwrap());
        let write = BufWriter::new(socket.try_clone().unwrap());
        Self { read, write }
    }
}

impl Read for BufTcpStream {
    #[inline]
    fn read(&mut self, buf: &mut [u8]) -> IoResult<usize> {
        self.read.read(buf)
    }

    #[inline]
    fn read_vectored(&mut self, bufs: &mut [IoSliceMut<'_>]) -> IoResult<usize> {
        self.read.read_vectored(bufs)
    }

    #[inline]
    fn read_to_end(&mut self, buf: &mut Vec<u8>) -> IoResult<usize> {
        self.read.read_to_end(buf)
    }

    #[inline]
    fn read_to_string(&mut self, buf: &mut String) -> IoResult<usize> {
        self.read.read_to_string(buf)
    }

    #[inline]
    fn read_exact(&mut self, buf: &mut [u8]) -> IoResult<()> {
        self.read.read_exact(buf)
    }
}

impl Write for BufTcpStream {
    #[inline]
    fn write(&mut self, buf: &[u8]) -> IoResult<usize> {
        self.write.write(buf)
    }

    #[inline]
    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> IoResult<usize> {
        self.write.write_vectored(bufs)
    }

    #[inline]
    fn flush(&mut self) -> IoResult<()> {
        self.write.flush()
    }

    #[inline]
    fn write_all(&mut self, buf: &[u8]) -> IoResult<()> {
        self.write.write_all(buf)
    }
}
// --- [/modification] ---

fn main() {
    let mut root_store = RootCertStore::empty();
    root_store.add_trust_anchors(webpki_roots::TLS_SERVER_ROOTS.iter().map(|ta| {
        OwnedTrustAnchor::from_subject_spki_name_constraints(
            ta.subject,
            ta.spki,
            ta.name_constraints,
        )
    }));
    let config = rustls::ClientConfig::builder()
        .with_safe_defaults()
        .with_root_certificates(root_store)
        .with_no_client_auth();

    let server_name = "www.rust-lang.org".try_into().unwrap();
    let mut conn = rustls::ClientConnection::new(Arc::new(config), server_name).unwrap();
    let sock = TcpStream::connect("www.rust-lang.org:443").unwrap();
    // --- [modification] ---
    let mut sock = BufTcpStream::new(&sock);
    // --- [/modification] ---
    let mut tls = rustls::Stream::new(&mut conn, &mut sock);
    tls.write_all( // <-- this call hangs until it times out
        concat!(
            "GET / HTTP/1.1\r\n",
            "Host: www.rust-lang.org\r\n",
            "Connection: close\r\n",
            "Accept-Encoding: identity\r\n",
            "\r\n"
        )
        .as_bytes(),
    )
    .unwrap();
    let ciphersuite = tls.conn.negotiated_cipher_suite().unwrap();
    writeln!(
        &mut std::io::stderr(),
        "Current ciphersuite: {:?}",
        ciphersuite.suite()
    )
    .unwrap();
    let mut plaintext = Vec::new();
    tls.read_to_end(&mut plaintext).unwrap();
    stdout().write_all(&plaintext).unwrap();
}
```

I also verified that the lack of flushing is indeed the culprit by changing 

https://github.com/rustls/rustls/blob/ac10422a898335230f64104b169461a50e9d61db/rustls/src/conn.rs#L383-L385

to  
```rust
while self.wants_write() {
    wrlen += self.write_tls(io)?;
}
io.flush()?;
```

which solves my issue.

I didn't submit a PR because I'm pretty sure there are other places that need flushing as well.

I'm using version `0.21.7`.

### PR DESCRIPTION
**Pull request by robsdedude**
Fixes #1537

### CODE REVIEW

File: rustls/src/conn.rs [hunk 1]
```diff
@@ -380,9 +380,13 @@ impl<Data> ConnectionCommon<Data> {
         loop {
             let until_handshaked = self.is_handshaking();
 
+            let wrlen_before = wrlen;
             while self.wants_write() {
                 wrlen += self.write_tls(io)?;
             }
+            if wrlen > wrlen_before {
```
[the remark this answers is not in the document]
Comment (ctz) [reply rustls/src/conn.rs:387]: I would go without this, and do the `flush()` unconditionally.

**Review by cpu — APPROVED** [review]
Thanks! This looks good to me. I also verified that backing out the changes in `rustls/src/conn.rs` but leaving the new unit tests provokes the expected failure as we'd expect :+1:

**Comment by cpu** [comment]
@robsdedude Would you mind squashing your commits down into one? That will let us do the usual merge queue w/ the rebase strategy while also keeping the commit history clean in `main`.

**Review by ctz — APPROVED** [review]
Thanks!

**Comment by cpu** [comment]
> Open question: should we backport this to 0.21.x?

I think we might as well. It seems like a clear cut bugfix and we have a branch in preparation already. I can add this.

**Pull request merged**